enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+- Modify the BuildController to use customized build pods instead of OpenShift Image Builder to create the finalized OS image (assuming we did the initial implementation with OpenShift Image Builder).
+- Add support for heterogeneous clusters using BuildController.
+
+#### Phase 3
+- oc command is augmented with sub-commands to provide syntactic sugar and a more polished synchronous user experience for cluster admins around staging a new OS image, watching preflight checks pass, and rolling it out. The web UI gains an equivalent UX flow.
+
+**Note:** Rest of the enhancement proposal will focus on. For other phases, details can be found in [Future Phases](#future-phases) section.
+
+### Workflow Description
+By default, cluster admin shouldn't see any change in the behavior unless they want to override default osImageURL.
+
+In order to apply a custom OS image, cluster admin would run following steps:
+1. Creates a custom OS image derived from default RHEL CoreOS container image
+1. Pushes this image to a container registry that is reachable to the cluster
+1. Creates a MachineConfig overriding osImageURL for targeted pool
+1. MCO will update OS with content provided via MachineConfig like it does today for any other MachineConfig being applied.
+1. In case of an error, error message will be made visible for further debugging
+
+### API Extensions
+MCO will allow `osImageURL` field to be overridden via the MachineConfig.
+
+### Implementation details
+
+#### Buildless layering support
+
+<!--Diagram Source: https://docs.google.com/drawings/d/1sM5QLxmAfx4GKZEM5Y5zNT-pT7t6shTkwdaM283-R3A/edit-->
+![Buildless layering support](buildless_layering_support.svg)
+
+In this phase, no final image build is required in MCO.
+Flow would remain same:
+1. Renders a new MachineConfig for desired pools.
+1. Performs cordon and drain operation on node.
+1. Writes rest of the MachineConfigs on the node
+1. Applies the new native container format RHEL CoreOS image to the node.
+1. Reboots the node
+
+#### Upgrade flow
+
+See [OSUpgrades](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md) for the existing upgrade behavior.
+
+- **With default osImageURL:**
+When cluster is not using any custom osImageURL, MCO will perform OS update as usual with what is supplied in the upcoming release image.
+- **With custom osImageURL:** With custom osImageURL provided, cluster admin grabs the steering wheel so they decide how and when to update the OS.
+  - MCO will respect the overridden osImageURL provided by cluster admin and will stay on the older image after upgrade.
+  - When cluster admin is ready to upgrade OS, they build new image derived from latest default RHEL CoreOS image supplied in release image and provides it via MachineConfig.
+  - At any point, the cluster can be switched back to default OS Image by deleting the custom osImageURL supplied via MachineConfig.
+
+#### Handling booting old nodes
```

Comment:
I feel like this is important enough to warrant an explicit call out in the phase 0 plans, to keep the ability of booting old nodes once on the new format

Reply:
umm, this whole section is talking mainly about phase 0. For even more clarity, will be adding phase 0 header explicitly in implementation details.